rts/Lua/LuaFonts.cpp

```diff
@@ -205,6 +207,49 @@ int LuaFonts::DeleteFont(lua_State* L)
 	return meta_gc(L);
 }
 
+/*** Adds a fallback font for the font rendering engine.
+ *
+ * Fonts added first will have higher priority.
+ * When a glyph isn't found when rendering a font, the fallback fonts
+ * will be searched first, otherwise os fonts will be used.
+ *
+ * The application should listen for the unsynced 'FontsChanged' callin so
+ * modules can clear any already reserved display lists or other relevant
+ * caches.
+ *
+ * Note the callin won't be executed at the time of calling this method,
+ * but later, on the Update cycle (before other Update and Draw callins).
+ *
+ * @function gl.AddFallbackFont
+ * @string filePath VFS path to the file, for example "fonts/myfont.ttf"
+ * @treturn bool success
+ */
+int LuaFonts::AddFallbackFont(lua_State* L)
+{
+	RECOIL_DETAILED_TRACY_ZONE;
+
+	const auto font = luaL_checkstring(L, 1);
+
+	const bool f = CFontTexture::AddFallbackFont(font);
```

Comment:
Add a string param for VFS mode and pass it along.
```suggestion
	const auto font = luaL_checkstring(L, 1);
	const auto vfsMode = luaL_optstring(L, 2, SPRING_VFS_RAW_FIRST);

	const bool f = CFontTexture::AddFallbackFont(font, vfsMode);
```

Reply:
Should we allow all vfs modes here? I guess since it's only loading fonts it's not too much of an issue, but I think vfs mode should be limited generally by the lua access level, I see the lua api allowing more vfs control in places where the vfs modes are restricted.

Also, other luafonts don't currently support this, so it's also going to be a bit of an api mismatch.

That said, I have no problem in adding support for this, but would like to be sure we're not allowing vfs modes that should be forbidden.

Reply:
ie, like here: https://github.com/beyond-all-reason/spring/blob/05c5724c3178571cb3bcdcc7d910dd6b13d6af21/rts/Lua/LuaVFS.cpp#L114-L122

Reply:
Synced doesn't have access to font interfaces, and a game can poison the call to block loading custom fonts from LuaUI if it wants to.

Reply:
I'm not convinced due to the api considerations:

- The other font interfaces don't have this parameter, so this would result in api mismatch.
  - Also probably means this hasn't been an issue till now for the rest of the api.
  - Maybe better to add this later for all the font api methods if we really want this, not just this one.
- Also, it's likely the api for AddFallbackFont will have to be extended in the future with new args. Namely I can see fallback fonts being registered for just some set of characters, or just for some main font. By adding the parameter now we will have to keep vfsmodes as the second parameter, imo not best since it's likely an option that won't be usually needed, so would be better to have after other more commonly used parameters. It's true this could be kind of overcome by allowing passing a table with all the options, but still not ideal from my pov.

Besides that, is there a real use case for this? I'm guessing there might be since you're asking for it, but since it's not something the rest of the api is doing generally, I have to ask. Please excuse me if it's something evident that I'm failing to see :D.

regarding the allowed modes:

> Synced doesn't have access to font interfaces, and a game can poison the call to block loading custom fonts from LuaUI if it wants to.

Right, I misread that to be unsynced who was getting more restricted. Ok, that's not an issue then.

Other than that, I already have this implemented (you basically provided the whole thing :)) in case we green light it, just let me know what you think of the above concerns.

Reply:
> The other font interfaces don't have this parameter, so this would result in api mismatch. 
> Maybe better to add this later for all the font api methods if we really want this, not just this one.

Yes, ideally everything that loads files would accept a VFS mode parameter. This applies even outside the Lua API (for example model textures that the engine loads automatically).

> Also probably means this hasn't been an issue till now for the rest of the api.
> is there a real use case for this?

I don't think it's a practical issue for fonts specifically, but I would like it to be consistent with everything else, and I would want everything else to eventually let you specify it because it is a problem there:
 * sometimes you want to hide something. For example sniper in ZK can shoot while invisible, which gives off particles that are purposefully blurry to let players know something is happening but requiring to pay attention and not giving away the location too much. Somebody can replace the blurry particle bitmaps with full-opacity, (255, 0, 0) red arrows and get much more info than the game dev intends. Ditto for other things like sounds. I don't mind the `VFS.RAW_FIRST` default on most other particles that exist to be pretty, but I would want to set `VFS.ZIP_ONLY` on the ones with gameplay consequences. Similar could apply to other content that carries information like sounds or shader textures.
 * sometimes you want exclusivity. For example custom skins/fx for contributors/donators. Here you don't want to allow getting the exclusive thing (if just locally) by merely copy-pasting the artwork under a different path.

> it's likely the api for AddFallbackFont will have to be extended in the future with new args. Namely I can see fallback fonts being registered for just some set of characters, or just for some main font. By adding the parameter now we will have to keep vfsmodes as the second parameter, imo not best since it's likely an option that won't be usually needed, so would be better to have after other more commonly used parameters. It's true this could be kind of overcome by allowing passing a table with all the options, but still not ideal from my pov.

Sure, it's not urgent and can definitely wait until later. Figuring a good way to pack both the path and the VFS mode into a single arg would be decent, but I'm sceptical whether good options exist given it's currently universally handled via an explicit mode param where available.

Reply:
> Sure, it's not urgent and can definitely wait until later. Figuring a good way to pack both the path and the VFS mode into a single arg would be decent, but I'm sceptical whether good options exist given it's currently universally handled via an explicit mode param where available.

I think this could be done through a lua userdata object (I think it's called like that) and some method to build it, like:

`local vfsPathWithMode = Spring.VFS.BuildPath(path, vfsModes)`

Then everywhere we could just check if it's a string or such userdata and act in consequence. This would be a good way since already existing apis could be adapted to take either a string with default vfs access or lua userdata with specific vfs access. Would also be backwards compatible.

Reply:
> I would like it to be consistent with everything else, and I would want everything else to eventually let you specify it because it is a problem there

Imo, since we already have several apis taking path as first parameter, with other parameters already going after that, I believe the way to do this from my previous comment would be the way to go, in order to keep the whole api consistent and backwards compatible.

I'd be happy to look into this in the near future.

Reply:
Userdata is some sort of desync fuel and is disallowed in synced. Table is probably alright. I've come to slightly dislike taking multiple different types of arguments (because it makes the function harder to document and people tend to copypaste the worst uses) but it's probably better than trying to cram extra data into strings. Anyway, changing the convention is outside the scope of this PR.

Reply:
> Userdata is some sort of desync fuel and is disallowed in synced. Table is probably alright. I've come to slightly dislike taking multiple different types of arguments (because it makes the function harder to document and people tend to copypaste the worst uses) but it's probably better than trying to cram extra data into strings. Anyway, changing the convention is outside the scope of this PR.

Yeah, I understand but don't really have more ideas. The fact the argument is a table or whatever can be hidden by recommending a method to build the object or creating an api that makes it look like you're actually passing an object so it feels prettier.

`gl.LoadFont(VFS.Path("fonts/blah.ttf"), 20, 3, 3)`

It's just a bit more unconvenient for api users when they need just a path and don't care about vfs mode, but that's all.

Also we could deprecate use of a string parameter. That way eventually there wouldn't be different types of arguments at the path spot, just the new system.

The alternative being add vfsmodes as nth parameter to every function currently taking a path parameter wouldn't be prettier, specially when later on more parameters are added and vfsmode is like the 4th parameter of 6 XD. Not even exaggerating: LoadFont for example currently has 3 arguments after path, and it's not unconceivable even more could be added later on.